toolchain/driver/runtimes_cache.cpp

```diff
@@ -0,0 +1,496 @@
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/FormatAdapters.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Support/SHA256.h"
+
+namespace Carbon {
+
+auto Runtimes::Make(std::filesystem::path path, llvm::raw_ostream* vlog_stream)
+    -> ErrorOr<Runtimes> {
+  if (!path.is_absolute()) {
+    return Error("Runtimes require an absolute path");
+  }
+
+  CARBON_ASSIGN_OR_RETURN(
+      Filesystem::Dir dir,
+      Filesystem::Cwd().OpenDir(path, Filesystem::OpenExisting));
+  return Runtimes(std::move(path), std::move(dir), {}, {}, vlog_stream);
+}
+
+auto Runtimes::Destroy() -> void {
+  // Release the lock on the runtimes and close the lock file.
+  flock_ = {};
+  auto close_result = std::move(lock_file_).Close();
+  if (!close_result.ok()) {
+    // Log and continue on close errors.
+    CARBON_VLOG("Error closing lock file for runtimes '{0}': {1}", base_path_,
+                close_result.error());
+  }
+}
+
+auto Runtimes::GetExistingClangResourceDir() -> ErrorOr<std::filesystem::path> {
+  std::filesystem::path path = base_path_ / ClangResourceDirPath;
+  auto open_result =
+      base_dir_.OpenDir(ClangResourceDirPath, Filesystem::OpenExisting);
+  if (open_result.ok()) {
+    return path;
+  }
+  return open_result.error().ToError();
+}
+
+auto Runtimes::BuildClangResourceDir()
+    -> ErrorOr<std::variant<std::filesystem::path, Builder>> {
+  return BuildClangResourceDirImpl(BuildLockDeadline, BuildLockPollInterval);
+}
+
+auto Runtimes::BuildClangResourceDirImpl(Filesystem::Duration deadline,
+                                         Filesystem::Duration poll_interval)
+    -> ErrorOr<std::variant<std::filesystem::path, Builder>> {
+  // Try to get an existing resource directory first.
+  auto existing_result = GetExistingClangResourceDir();
+  if (existing_result.ok()) {
+    return {*std::move(existing_result)};
+  }
+
+  // Otherwise, we will need to build the runtimes and commit them into this
+  // directory once ready. Try and acquire an advisory lock to avoid redundant
+  // computation.
+  CARBON_ASSIGN_OR_RETURN(
+      Filesystem::ReadWriteFile lock_file,
+      base_dir_.OpenReadWrite(
+          llvm::formatv(LockFileFormat, ClangResourceDirPath).str(),
+          Filesystem::OpenAlways, /*creation_mode=*/0700));
+  CARBON_VLOG("PID {0} locking cache path: {1}\n", getpid(),
+              base_path_ / ClangResourceDirPath);
+  Filesystem::FileLock flock;
+  auto flock_result = lock_file.TryLock(Filesystem::FileLock::Exclusive,
+                                        deadline, poll_interval);
+  if (flock_result.ok()) {
+    flock = *std::move(flock_result);
+    CARBON_VLOG("Successfully locked cache path\n");
+    // As a debugging aid, write our PID into the lock file when we
+    // successfully acquire it. Ignore errors here though.
+    auto write_result = lock_file.WriteFileFromString(std::to_string(getpid()));
+    static_cast<void>(write_result);
```

Comment:
Optional:

```suggestion
    (void)write_result;
```

Both forms seem to occur in our codebase with about the same frequency, but I find this one easier to read. With that change, it might also make sense to drop `write_result` and discard the function return value directly.

Reply:
Sure, done.